crates/wasi/src/preview2/preview2/filesystem.rs

```diff
@@ -521,31 +556,58 @@ impl<T: WasiView> filesystem::Host for T {
                 Err(ErrorCode::Invalid)?;
             }
         }
-        let mut opened = d.dir.open_with(&path, &opts)?;
 
-        if opened.metadata()?.is_dir() {
-            Ok(table.push_dir(Dir::new(
-                cap_std::fs::Dir::from_std_file(opened.into_std()),
-                d.perms,
-                d.file_perms,
-            ))?)
-        } else if oflags.contains(OpenFlags::DIRECTORY) {
-            Err(ErrorCode::NotDirectory)?
-        } else {
-            // FIXME cap-std needs a nonblocking open option so that files reads and writes
-            // are nonblocking. Instead we set it after opening here:
-            let set_fd_flags = opened.new_set_fd_flags(FdFlags::NONBLOCK)?;
-            opened.set_fd_flags(set_fd_flags)?;
+        // Represents each possible outcome from the spawn_blocking operation.
+        // This makes sure we don't have to give spawn_blocking any way to
+        // manipulate the table.
+        enum OpenResult {
+            Dir(cap_std::fs::Dir),
+            File(cap_std::fs::File),
+            NotDir,
+        }
+
+        let opened = d
+            .spawn_blocking::<_, std::io::Result<OpenResult>>(move |d| {
+                let mut opened = d.open_with(&path, &opts)?;
+                if opened.metadata()?.is_dir() {
+                    Ok(OpenResult::Dir(cap_std::fs::Dir::from_std_file(
+                        opened.into_std(),
+                    )))
+                } else if oflags.contains(OpenFlags::DIRECTORY) {
+                    Ok(OpenResult::NotDir)
+                } else {
+                    // FIXME cap-std needs a nonblocking open option so that files reads and writes
+                    // are nonblocking. Instead we set it after opening here:
+                    let set_fd_flags = opened.new_set_fd_flags(FdFlags::NONBLOCK)?;
+                    opened.set_fd_flags(set_fd_flags)?;
+                    Ok(OpenResult::File(opened))
+                }
+            })
+            .await?;
+
+        match opened {
+            OpenResult::Dir(dir) => Ok(table.push_dir(Dir::new(dir, d.perms, d.file_perms))?),
+
+            OpenResult::File(file) => {
+                Ok(table.push_file(File::new(file, mask_file_perms(d.file_perms, flags)))?)
+            }
 
-            Ok(table.push_file(File::new(opened, mask_file_perms(d.file_perms, flags)))?)
+            OpenResult::NotDir => Err(ErrorCode::NotDirectory.into()),
         }
     }
 
     async fn drop_descriptor(&mut self, fd: filesystem::Descriptor) -> anyhow::Result<()> {
         let table = self.table_mut();
+
+        // The Drop will close the file/dir, but if the close syscall
+        // blocks the thread, I will face god and walk backwards into hell.
+        // tokio::fs::File just uses std::fs::File's Drop impl to close, so
+        // it doesn't appear anyone else has found this to be a problem.
+        // (Not that they could solve it without async drop...)
```

Comment:
Had a good chuckle reading this :)

Regardless I agree we can't make this truly async and sync close here is the right thing to do IMO